File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            // No change is needed; return the current head
            return head;
        }
        
        ListNode* prev = NULL;   // Initialize previous pointer to NULL
        ListNode* current = head; // Start with the head node
        
        while (current != NULL) {
            ListNode* nextNode = current->next; // Temporarily store the next node
            current->next = prev;              // Reverse the link
            prev = current;                    // Move prev to current
            current = nextNode;                // Move to the next node
        }

        return prev; // New head of the reversed list
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {

        ListNode prev=null;
        ListNode mid=null;
        ListNode curr=head;
    
        while(curr!=null){
        //   ListNode next=curr.next;
        //     curr.next=prev;
        //     prev=curr;
        //     curr=next;

        prev=mid;
        mid=curr;
        curr=curr.next;
        mid.next=prev;


        }
        
        //   return prev; 
        return mid;
    }
}